### ISSUE TITLE
File upload field failing when variable used in subpath

### ISSUE THREAD

**Issue by jrrdnx** [barrelstrength/craft-sprout-forms#439]
I'm attempting to segment file uploads into their own directories for each submission, but the File Upload field is failing to upload a file when using the variable `{slug}` in the subfolder path.

### Additional info

- Craft version: 3.4.8
- PHP version: 7.4.2
- Database driver & version:  MySQL 5.7.28
- Plugins & versions: Amazon S3 1.2.7
Sprout Forms 3.8.4

Applicable section from web.log:

```
2020-02-27 10:06:46 [-][-][8g86vrr3qscgatdkllhti3m8d7][error][barrelstrength\sproutforms\services\Entries::saveEntry] Failed to save element: The Upload Documents field’s Upload Location setting has an invalid subpath (“{slug}”).
2020-02-27 10:06:46 [-][-][8g86vrr3qscgatdkllhti3m8d7][error][craft\errors\InvalidSubpathException] craft\errors\InvalidSubpathException: Could not resolve the subpath “{slug}”. in /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/fields/formfields/FileUpload.php:799
Stack trace:
#0 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/fields/formfields/FileUpload.php(890): barrelstrength\sproutforms\fields\formfields\FileUpload->_resolveVolumePathToFolderId()
#1 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/fields/formfields/FileUpload.php(491): barrelstrength\sproutforms\fields\formfields\FileUpload->_determineUploadFolderId()
#2 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/fields/formfields/FileUpload.php(500): barrelstrength\sproutforms\fields\formfields\FileUpload->barrelstrength\sproutforms\fields\formfields\{closure}()
#3 /home/domain.com/domain.com/vendor/craftcms/cms/src/base/Element.php(2529): barrelstrength\sproutforms\fields\formfields\FileUpload->afterElementSave()
#4 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/elements/Entry.php(472): craft\base\Element->afterSave()
#5 /home/domain.com/domain.com/vendor/craftcms/cms/src/services/Elements.php(2123): barrelstrength\sproutforms\elements\Entry->afterSave()
#6 /home/domain.com/domain.com/vendor/craftcms/cms/src/services/Elements.php(530): craft\services\Elements->_saveElementInternal()
#7 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/services/Entries.php(167): craft\services\Elements->saveElement()
#8 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/controllers/EntriesController.php(256): barrelstrength\sproutforms\services\Entries->saveEntry()
#9 [internal function]: barrelstrength\sproutforms\controllers\EntriesController->actionSaveEntry()
#10 /home/domain.com/domain.com/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array()
#11 /home/domain.com/domain.com/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams()
#12 /home/domain.com/domain.com/vendor/craftcms/cms/src/web/Controller.php(178): yii\base\Controller->runAction()
#13 /home/domain.com/domain.com/vendor/yiisoft/yii2/base/Module.php(528): craft\web\Controller->runAction()
#14 /home/domain.com/domain.com/vendor/craftcms/cms/src/web/Application.php(291): yii\base\Module->runAction()
#15 /home/domain.com/domain.com/vendor/craftcms/cms/src/web/Application.php(559): craft\web\Application->runAction()
#16 /home/domain.com/domain.com/vendor/craftcms/cms/src/web/Application.php(270): craft\web\Application->_processActionRequest()
#17 /home/domain.com/domain.com/vendor/yiisoft/yii2/base/Application.php(386): craft\web\Application->handleRequest()
#18 /home/domain.com/domain.com/web/index.php(24): yii\base\Application->run()
#19 {main}

Next craft\errors\InvalidSubpathException: The Upload Documents field’s Upload Location setting has an invalid subpath (“{slug}”). in /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/fields/formfields/FileUpload.php:909
Stack trace:
#0 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/fields/formfields/FileUpload.php(491): barrelstrength\sproutforms\fields\formfields\FileUpload->_determineUploadFolderId()
#1 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/fields/formfields/FileUpload.php(500): barrelstrength\sproutforms\fields\formfields\FileUpload->barrelstrength\sproutforms\fields\formfields\{closure}()
#2 /home/domain.com/domain.com/vendor/craftcms/cms/src/base/Element.php(2529): barrelstrength\sproutforms\fields\formfields\FileUpload->afterElementSave()
#3 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/elements/Entry.php(472): craft\base\Element->afterSave()
#4 /home/domain.com/domain.com/vendor/craftcms/cms/src/services/Elements.php(2123): barrelstrength\sproutforms\elements\Entry->afterSave()
#5 /home/domain.com/domain.com/vendor/craftcms/cms/src/services/Elements.php(530): craft\services\Elements->_saveElementInternal()
#6 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/services/Entries.php(167): craft\services\Elements->saveElement()
#7 /home/domain.com/domain.com/vendor/barrelstrength/sprout-forms/src/controllers/EntriesController.php(256): barrelstrength\sproutforms\services\Entries->saveEntry()
#8 [internal function]: barrelstrength\sproutforms\controllers\EntriesController->actionSaveEntry()
#9 /home/domain.com/domain.com/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array()
#10 /home/domain.com/domain.com/vendor/yiisoft/yii2/base/Controller.php(157): yii\base\InlineAction->runWithParams()
#11 /home/domain.com/domain.com/vendor/craftcms/cms/src/web/Controller.php(178): yii\base\Controller->runAction()
#12 /home/domain.com/domain.com/vendor/yiisoft/yii2/base/Module.php(528): craft\web\Controller->runAction()
#13 /home/domain.com/domain.com/vendor/craftcms/cms/src/web/Application.php(291): yii\base\Module->runAction()
#14 /home/domain.com/domain.com/vendor/craftcms/cms/src/web/Application.php(559): craft\web\Application->runAction()
#15 /home/domain.com/domain.com/vendor/craftcms/cms/src/web/Application.php(270): craft\web\Application->_processActionRequest()
#16 /home/domain.com/domain.com/vendor/yiisoft/yii2/base/Application.php(386): craft\web\Application->handleRequest()
#17 /home/domain.com/domain.com/web/index.php(24): yii\base\Application->run()
#18 {main}
2020-02-27 10:06:45 [-][-][8g86vrr3qscgatdkllhti3m8d7][info][application] $_GET = []
```

<img src="https://frontapp.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_1fbpf)

**Comment by BenParizek** [comment]
Thanks for reporting @jrrdnx. I can recreate this and will need to look into this. The Relations fields in Sprout Forms piggyback on Craft's Relations fields a bit so I expect we're not adjusting an assumption somewhere. I'll update this thread when I have an update.

**Comment by BenParizek** [comment]
@jrrdnx This behavior has been fixed in Sprout Forms v3.8.8.

I've also updated the instructions text to be more form specific. Please note the previous text was applicable to the Craft Asset Field use case and values like `{slug}` are only available when the Asset Field is processing an Entry Model that has the `{slug}` variable. 

In the Sprout Forms use case, when the File Upload field gets processed the submitted Form Entry Model is passed to it so you'll need to use attributes available to that model. As an example, `{id}` or `{{ object.id }} refer to the `id` on the Form Entry Model and if you want to use some value from the Form that the Form Entry Model relates to, you can access the Form using `{{ object.getForm().id }}`. I use `id` here just because it's simple to share. You can use any available attributes and custom fields in your dynamic settings.
